File: Behavioural-Patterns/Visitor/Java/Implementations/WorkflowValidatorVisitor.java
package Implementations;

import Interfaces.WorkflowVisitor;

public class WorkflowValidatorVisitor implements WorkflowVisitor {

    @Override
    public void visit(EmailTriggerStep emailTriggerStep) {
        System.out.print("Validating Email Trigger... ");
        if (emailTriggerStep.getMailbox() == null || emailTriggerStep.getMailbox().isEmpty()) {
            System.out.println("ERROR: Mailbox is not configured.");
        } else {
            System.out.println("OK.");
        }
    }

    @Override
    public void visit(FilterStep filterStep) {
        System.out.print("Validating Filter... ");
        if (filterStep.getCondition() == null || filterStep.getCondition().isEmpty()) {
            System.out.println("ERROR: Filter condition is empty.");
        } else {
            System.out.println("OK.");
        }
    }

    @Override
    public void visit(CreateTaskStep createTaskStep) {
        System.out.print("Validating Create Task Step... ");
        if (createTaskStep.getProjectName() == null || createTaskStep.getProjectName().isEmpty()) {
            System.out.println("ERROR: Project name is missing.");
        } else {
            System.out.println("OK.");
        }
    }

    @Override
    public void visit(SlackNotificationStep slackNotificationStep) {
        System.out.print("Validating Slack Step... ");
        if (slackNotificationStep.getChannel() == null || slackNotificationStep.getChannel().isEmpty()) {
            System.out.println("ERROR: Slack channel is not configured.");
        } else {
            System.out.println("OK.");
        }
    }
    
}

File: Behavioural-Patterns/Visitor/C#/Visitor/Implementations/WorkflowValidatorVisitor.cs
public class WorkflowValidatorVisitor : WorkflowVisitor {

    public void visit(EmailTriggerStep emailTriggerStep) {
        Console.WriteLine("Validating Email Trigger... ");
        if (emailTriggerStep.getMailbox() == null || emailTriggerStep.getMailbox().Length == 0) {
            Console.WriteLine("ERROR: Mailbox is not configured.");
        } else {
            Console.WriteLine("OK.");
        }
    }

    public void visit(FilterStep filterStep) {
        Console.WriteLine("Validating Filter... ");
        if (filterStep.getCondition() == null || filterStep.getCondition().Length == 0) {
            Console.WriteLine("ERROR: Filter condition is empty.");
        } else {
            Console.WriteLine("OK.");
        }
    }

    public void visit(CreateTaskStep createTaskStep) {
        Console.WriteLine("Validating Create Task Step... ");
        if (createTaskStep.getProjectName() == null || createTaskStep.getProjectName().Length == 0) {
            Console.WriteLine("ERROR: Project name is missing.");
        } else {
            Console.WriteLine("OK.");
        }
    }

    public void visit(SlackNotificationStep slackNotificationStep) {
        Console.WriteLine("Validating Slack Step... ");
        if (slackNotificationStep.getChannel() == null || slackNotificationStep.getChannel().Length == 0) {
            Console.WriteLine("ERROR: Slack channel is not configured.");
        } else {
            Console.WriteLine("OK.");
        }
    }
    
}
